Restructure the documentation for custom images (#312)
rancher/elemental-docs: docs/custom-images.md

---
sidebar_label: Build Custom OS Images
title: ''
---

<head>
  <link rel="canonical" href="https://elemental.docs.rancher.com/custom-images"/>
</head>

# How to build and use custom OS images

## Remasterimg an OS image with a custom Dockerfile

> Since OS images provided by Elemental are Docker images, they can also be used as a base image

Reviewer: ```suggestion
Since OS images provided by Elemental are container images, they can also be used as a base image
```

in a Dockerfile in order to create a new container image.

> Imagine some additional package from an extra repository is required, the following example

Reviewer 2: ```suggestion
Imagine some additional packages from an extra repository is required, the following example
```

> show case how this could be added:

Reviewer 2: ```suggestion
showcases how this could be added:
```

RUN elemental build-iso \
        dir:rootfs \
        --bootloader-in-rootfs \
        --squash-no-compression \
        -o /output -n "elemental-${TARGETARCH}"

FROM busybox
COPY --from=builder /output /custom-iso

ENTRYPOINT ["busybox", "sh", "-c"]
```

Build it with regular `docker build` command:

```bash showLineNumbers
docker build -t myrepo/custom-build:v1.1.1 \
              --build-arg IMAGE_REPO=myrepo/custom-build-iso \
              --build-arg IMAGE_TAG=v1.1.1 \
              .
```

The resulting container image is actually a container image including the ISO,
this container image can be pushed to an OCI registry too. The ISO image can be
extracted from the container to the current folder by executing the container as:

```bash showLineNumbers
docker run --rm -v $(pwd):/host mytest-image "busybox cp /custom-iso/*.iso /host"
```

The new customized installation media can be found in `elemental-<arch>.iso`.

The above container run is equivalent to what *elemental-operator* does to extract
the ISO from a container to build a new one including the registration URL,
hence this is also a good check mark to verify the container can be pushed to a

> registry and used in *elemental-operator* as a `baseImage` for a

Reviewer 2: I see why you used "in" here... anyway, seems to me sounds better (and may be more clear to the casual user) if we use "by" here...
btw, drop the suggestion if you prefer "in".
```suggestion
registry and used by the *elemental-operator* as a `baseImage` for a
```

[SeedImage](seedimage-reference) resource.

## List custom images as a ManagedOSVersion resource

In Elemental listing OS container images and ISO container images as ManagedOSVersion
resources is not mandatory but handy. Specially from a UI perspective this makes
the custom images visible and easy to use from the Elemental UI extension.

Continuing the example from the previous section a custom OS container referenced as

> `myrepo/custom-build:v1.1.1` was build and eventually pushed to a registry. Then this

Reviewer 2: ```suggestion
`myrepo/custom-build:v1.1.1` was built and eventually pushed to a registry. Then this
```

> image ready to be added as a ManagedOSVersion resource with:

Reviewer 2: ```suggestion
image is ready to be added as a ManagedOSVersion resource with:
```
